docs/proposals/20191017-control-plane.md

```diff
@@ -0,0 +1,494 @@
+}
+
+// +kubebuilder:object:root=true
+
+// ControlPlaneList contains a list of ControlPlane
+type ControlPlaneList struct {
+    metav1.TypeMeta `json:",inline"`
+    metav1.ListMeta `json:"metadata,omitempty"`
+    Items           []ControlPlane `json:"items"`
+}
+
+func init() {
+    SchemeBuilder.Register(&ControlPlane{}, &ControlPlaneList{})
+}
+
+```
+
+With the following validations:
+
+- If the referenced InitTemplate does not define external etcd (webhook):
+  - ControlPlane.Spec.Replicas in [1, 3, 5, 7]
+- ControlPlane.Spec.Version != "" (openapi)
+
+And the following defaulting:
+
+- ControlPlaneSpec.Replicas: 1 (openapi or webhook)
+
+#### Modifications required to existing API Types
+
+- Add Cluster.Spec.ControlPlaneRef defined as:
+
+```go
+    // ControlPlaneRef is an optional reference to a provider-specific resource that holds
+    // the details for provisioning the Control Plane for a Cluster
+    // +optional
+    ControlPlaneRef *corev1.ObjectReference `json:"controlPlaneRef,omitempty"`
+```
+
+- Add Cluster.Status.ControlPlaneReady defined as:
+
+```go
+    // ControlPlaneReady defines if the control plane is ready
+    // +optional
+    ControlPlaneReady bool `json:"controlPlaneReady"`
+```
+
+#### Behavioral Changes from v1alpha2
+
+- If Cluster.Spec.ControlPlaneRef is set:
+  - [Status.ControlPlaneInitialized](https://github.com/kubernetes-sigs/cluster-api/issues/1243) is set based on the value of Status.Initialized for the referenced resource
+  - Status.ControlPlaneReady is set based on the value of Status.Ready for the referenced resource, this field is intended to eventually replace Status.ControlPlaneInitialized as a field that will be kept up to date instead of set only once.
+- Current behavior will be preserved if Cluster.Spec.ControlPlaneRef is not set
+
+#### Behaviors
+
+##### Create
+
+- A created ControlPlane should bootstrap a control plane with a given number of replicas
+- If an error occurs, errorStatus and errorReason will be defined on the ControlPlane object
+- Cannot create an even number of stacked control planes or anything other than 1, 3, 5, or 7
+- Can create an arbitrary number of control planes if etcd is external to the control plane
+- Creating a ControlPlane with > 1 replicas is equivalent to creating a ControlPlane with 1 replica followed by scaling the ControlPlane to 3 replicas
+- The kubeadm bootstrapping configuration provided via ControlPlane.Spec.KubeadmConfigSpec, should specify the initConfiguration, clusterConfiguration, and joinConfiguration stanzas, and the ControlPlane controller will be responsible for splitting the config and passing it to the underlying Machines created as appropriate
+- The underlying query used to find existing Control Plane Machines is based on the following hardcoded label selector:
+
+```yaml
+  selector:
+    matchLabels:
+      cluster.x-k8s.io/cluster-name: my-cluster
+      cluster.x-k8s.io/controlplane: "true"
```

Comment:
Can you describe a time when there'd be a machine with this label but it's not owned by CAPI?

Reply:
What if users might want to manage control plane machines via external bootstrap methods? Or as a single machines like in v1alpha2?

Reply:
Why? If someone is running a manual control plane today they wouldn't have a ControlPlane resource defined. It seems like adding an additional `owned` label would just complicate the adoption scenario.

Reply:
If they are creating actual `Machine` instances to be used as control plane members, starting in v1a3 they must use the `ControlPlane` to do so. We will no longer support other methods. If they want to manage the control plane outside of CAPI (e.g. `kops` phased adoption), that is totally fine and isn't impacted by this.

Reply:
This assumption and breaking change (which I'm ok with) should be clear in the proposal, is it outlined anywhere?

In addition, the current proposal doesn't support AZ-aware control planes, as outlined below. If we don't solve that use case, I think we shouldn't disallow to use plain Machine objects as control planes.

Reply:
Maybe my `must` is too strong? That's my preference, at least. But I recognize it might be biting off too much for alpha3.

Reply:
Yeah, I'm trying to understand the expectations. 

One clarification about my older comment, by "external bootstrap" I meant other bootstrap providers not living in CAPI that generate `Bootstrap.Data`. The `cluster.x-k8s.io/controlplane: "true"` label we're re-using here, is used to allow proper deletion for example, which we might need to retain.

Reply:
If they are using a different bootstrapper, I wouldn't expect them to use the default implementation of ControlPlane (or they would use an implementation of ControlPlane that works with their bootstrapper).

For AZ aware control planes I would expect us to lean on the proposal around representing failure domains rather than try to define it in this proposal: https://github.com/kubernetes-sigs/cluster-api/issues/46

I'll work on getting that incorporated into the doc later.

Reply:
> If they are using a different bootstrapper, I wouldn't expect them to use the default implementation of ControlPlane (or they would use an implementation of ControlPlane that works with their bootstrapper).

That's fine, but the use of the label still applies, which we expect to allow for proper deletion and control plane init.

> For AZ aware control planes I would expect us to lean on the proposal around representing failure domains rather than try to define it in this proposal: #46

Sounds good to me, thanks!

Reply:
> If they are using a different bootstrapper, I wouldn't expect them to use the default implementation of ControlPlane 

Based on the discussion above I'm concerned we're starting to create a series of disparate components (control plane controller, bootstrapped controller) that all work together well when using the reference implementations of each, but start to fall down when we start picking and replacing one or the other. 

Hmmm. I wonder if we should consider bundling these providers under a single aegis. Not just repo, but making them part of the CAPI controller manager and making it a conscious choice to disable them when configuring CAPI.

Reply:
CABPK is already part of the CAPI controller manager for v1alpha3, this controller will also be as well.

That said, you are raising a concern that I have around the interoperability of components when "swapping the batteries" since we will now be having things under the same tree.

Reply:
> CABPK is already part of the CAPI controller manager for v1alpha3

Got it.

> That said, you are raising a concern that I have around the interoperability of components when "swapping the batteries" since we will now be having things under the same tree.

Exactly. I don't really see any value in these being modular components if they're mutually modular. At some point the value of modularity seems to only what we're saying, not how we're actually implementing things.

Reply:
I believe there is still value in this modularity. And it's ok that we have default implementations for control plane and bootstrapper that go together. The key value proposition is the generic linkage from `Cluster` to some control plane resource via `.spec.controlPlaneRef`, and from `Machine` to some bootstrap config resource via `.spec.bootstrap.configRef`. These are not prescriptive, and the contracts we've defined and are defining for control plane and bootstrap resources still hold (e.g. they must have `.status.ready`, `.status.errorReason`, and the like).

I am still optimistic that we'll see other bootstrap and control plane providers and that they'll fit in nicely with the aforementioned contracts.

Reply:
> If they are using a different bootstrapper, I wouldn't expect them to use the default implementation of ControlPlane (or they would use an implementation of ControlPlane that works with their bootstrapper).

@detiber Sorry if I'm asking an obvietey but, what exactly would prevent this to work? What specific assumptions are made about the bootstrapper? I understand it it kubeadm bootstrapping configuration.

Reply:
> @detiber Sorry if I'm asking an obvietey but, what exactly would prevent this to work? What specific assumptions are made about the bootstrapper? I understand it it kubeadm bootstrapping configuration.

In addition to the embedded types that are specific to the Kubeadm bootstrapper, there are going to be assumptions made in the ways that the controller interacts with the components based on the way that kubeadm configures clusters. 

For example, the way that we'll manage etcd will be executing commands through the api server against pods named using the conventions of kubeadm, and referencing file paths that are used by kubeadm. We'll also be modifying a ConfigMap that is added and managed by kubeadm that is referenced when joining additional control plane nodes. As we continue to define the health checks, they will be highly dependent on the way that kubeadm configures it's static pods.

Reply:
I'm going to defer adding an `owned` label for now and we can revisit as part of the implementation if needed.